Enhance AIX shared library build to use an export List.

KamathForAIX opened the pull request:
Teach Meson how to create an export file in AIX.

KamathForAIX committed:
Enhance AIX shared library build to use an export List.

Teach Meson to create an export file in AIX for a shared library target.

KamathForAIX commented:
Hi @eli-schwartz and community members,

I had to close https://github.com/mesonbuild/meson/pull/14335 due to merge conflicts and changes in ninja-backend.py file coming which were not made by me.

Apologies for the same.

jpakkane commented on mesonbuild/scripts/symbolextractor.py:
```diff
@@ -122,6 +123,21 @@ def solaris_syms(libfilename: str, outfilename: str) -> None:
         os.environ['PATH'] = origpath
 
+def aix_syms(libfilename: str, outfilename: str) -> None:
+    nm_output = subprocess.run(rf"nm -BCg {libfilename} | egrep ' [TDB] ' | sed -e 's/.* //' | egrep -v '\\$' | sed -e 's/^[.]//' | sort | uniq",
```
Shell pipelines are *never* acceptable. This same functionality _must_ be provided with Python.

KamathForAIX replied:
Sure. Thanks, @jpakkane. I will definitely try to change this and will keep it in my mind. As of now, we are redesigning this pull request with @eli-schwartz. Post that I will try. 

In case of any constraints will get back.

KamathForAIX replied:
Thanks @bonzini for the feedback.

I am currently resolving the merge conflicts in this pull request and retesting.

Will let you know once I am done.

bonzini commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -3574,6 +3594,10 @@ def get_target_type_link_args(self, target: build.BuildTarget, linker: Compiler)
+            # Add export file option in command for AIX.
+            if self.environment.machines[target.for_machine].is_aix():
+                if not isinstance(target, build.SharedModule):
+                    commands += ['-Wl,-bE:' + self.get_target_filename(target) + '.exp']
```
Do not hardcode this. Instead add functions to the StaticLinker and DynamicLinker, such as `exported_symbols_args(self, target: BuildTarget, dir: str)`. It can be overridden to return `self._apply_prefix('-bE:' + os.path.join(dir, ...))` for the AIX linker.

Compiler then can add a matching `get_exported_symbols_args` function that calls the method for the linker, and that's what `get_target_type_link_args` will call.

Also, please place the `.exp` file in the target-private directory.

KamathForAIX replied:
This is a nice idea. I am trying this. Thanks.

bonzini commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -1111,6 +1098,48 @@ def generate_target(self, target: T.Union[build.BuildTarget, build.CustomTarget,
+        # /lib/libpthreads.a(shr_xpg5_64.o), /libpq/libpq.a(libpq.so.5.18) and so on. We need to tell the linker
+        # via the export file to add the symbols exported by libpq.a to libpq.a's symbol table, so at runtime linking
+        # the linker knows where these symbols came from and can resolve the same. The below code does two things.
+        # One with ld -r -o <name>_SUBSYS.o <object files> we create a single object file that will have all the
```
Meson should already have functionality to do this already, see `generate_prelink`.

Please make sure that prelinking works for AIX static libraries and build on top of that.

KamathForAIX replied:
Yes, Meson does. Understanding the code around me, I found something else also.

https://github.com/mesonbuild/meson/blob/master/mesonbuild/backend/ninjabackend.py#L3521

This function generate_shsym() here will do for AIX. What we need to do is after the shared library object is created and just before the archive is made, we need to complete the symbol extraction and then pass the symbol file via -Wl,-bE:<symbol_file> to the archive.

bonzini replied:
I wasn't sure about the format of the symbols file. If it doesn't match you can create e second one for the linker.

bonzini commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -1111,6 +1098,48 @@ def generate_target(self, target: T.Union[build.BuildTarget, build.CustomTarget,
+        # option which tells the linker that this shared library exports these symbols in export file.
+        # Example:
+        # build /getpwuid-preload.so_SUBSYS.o: Export_List_AIX_Object /getpwuid-preload.so.p/getpwuid-preload.c.o
+        # IMPLIB = /getpwuid-preload.a
```
IMPLIB is unnecessary here, since it is not referred to by the rule.

bonzini commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -1111,6 +1098,48 @@ def generate_target(self, target: T.Union[build.BuildTarget, build.CustomTarget,
+        symname = ""
+        if (isinstance(target, build.SharedLibrary) and not isinstance(target, build.SharedModule)
+                and self.environment.machines[target.for_machine].is_aix()):
+            target_file = os.path.join(self.get_target_dir(target), target.get_filename() + '_SUBSYS.o')
```
Please move this to a separate function, like

```diff
diff --git a/mesonbuild/backend/ninjabackend.py b/mesonbuild/backend/ninjabackend.py
index 7bf837bcc..3011aaab0 100644
--- a/mesonbuild/backend/ninjabackend.py
+++ b/mesonbuild/backend/ninjabackend.py
@@ -1104,10 +1104,12 @@ class NinjaBackend(backends.Backend):
             # Skip the link stage for this special type of target
             return
 
-        if not isinstance(target, build.StaticLibrary):
-            final_obj_list = obj_list
-        elif target.prelink:
+        if isinstance(target, build.StaticLibrary) and target.prelink:
             final_obj_list = self.generate_prelink(target, obj_list)
+        elif (self.environment.machines[target.for_machine].is_aix() and
+              isinstance(target, build.SharedLibrary) and
+              not isinstance(target, build.SharedModule)):
+            final_obj_list = self.generate_aix_prelink_and_export_list(target, obj_list)
         else:
             final_obj_list = obj_list
 
```

bonzini reviewed (changes requested):
I left some comments on how to reuse existing functionality and avoid hardcoding AIX-specific functionality.

KamathForAIX commented:
@bonzini Thank you for the review. I am on this. Just cleared the merge conflicts. I have issues in the AIX ecosystem while building. Will resolve and push soon.

bonzini commented:
More comments [here](https://github.com/mesonbuild/meson/commit/a6d8e49f529aac9a1f03adab84fe0bd64601c653).

KamathForAIX commented:
@bonzini

Please let me know about this version of the patch.

I am doing a git fetch and git rebase to remove the sync commit. 

Here is the log of how this version is designed to work.
```
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: /opt/freeware/bin/ninja -C /test_meson_project/build -v
ninja: Entering directory `/test_meson_project/build'
[1/7] gcc -Imyapp.p -I. -I.. -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -Wpedantic -O0 -g -MD -MQ myapp.p/main.c.o -MF myapp.p/main.c.o.d -o myapp.p/main.c.o -c ../main.c
[2/7] gcc -Ilibmylib.so.1.0.0.p -I. -I.. -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -Wpedantic -O0 -g -fPIC -MD -MQ libmylib.so.1.0.0.p/mylib.c.o -MF libmylib.so.1.0.0.p/mylib.c.o.d -o libmylib.so.1.0.0.p/mylib.c.o -c ../mylib.c
[3/7] /usr/bin/ld -r -o libmylib.so.1.0.0_COMBINED_OBJ.o libmylib.so.1.0.0.p/mylib.c.o
[4/7] /meson/install/opt/freeware/bin/meson --internal symbolextractor /test_meson_project/build libmylib.so.1.0.0_COMBINED_OBJ.o libmylib.a libmylib.so.1.0.0.p/libmylib.so.1.0.0.symbols 
[5/7] gcc  -o libmylib.so.1.0.0 libmylib.so.1.0.0.p/mylib.c.o -Wl,-bernotok -Wl,-bnoipath -Wl,-bbigtoc -shared -fPIC -Wl,-bE:/test_meson_project/build/libmylib.so.1.0.0.p/libmylib.so.1.0.0.symbols -Wl,-blibpath:/opt/freeware/lib/gcc/powerpc-ibm-aix7.2.0.0/13:/opt/freeware/lib:/lib:/usr/lib
[6/7] ar -r -s -v libmylib.a libmylib.so.1.0.0 && rm -f libmylib.so.1.0.0
ar: Creating an archive file libmylib.a.
a - libmylib.so.1.0.0
[7/7] gcc  -o myapp myapp.p/main.c.o -Wl,-bernotok -Wl,-bnoipath -Wl,-bbigtoc -Wl,-blibpath:/test_meson_project/build/:/opt/freeware/lib/gcc/powerpc-ibm-aix7.2.0.0/13:/opt/freeware/lib:/lib:/usr/lib libmylib.a
```

Below is the meson.build file

```
# cat meson.build
project('aix_export_test', 'c',
  version : '1.0.0',
  default_options : ['warning_level=3'])

lib_sources = ['mylib.c']

mylib = shared_library('mylib',
  lib_sources,
  install : true,
  version : '1.0.0',
  soversion : '1')

exe = executable('myapp',
  'main.c',
  link_with : mylib,
  install : true)

test('basic', exe)
```

[build.ninja.c](https://github.com/user-attachments/files/27197585/build.ninja.c)

KamathForAIX commented:
@bonzini 

I have also attached a build.ninja file for everyone to see the approach/logic. Had to name it .c otherwise it was not allowing me to attach.

bonzini commented:
The main remaining comment (other than the various CI failures) is why not use prelinking. Apart from that I am not sure asking nm to demangle is correct.

KamathForAIX commented:
@bonzini 

> The main remaining comment (other than the various CI failures) is why not use prelinking. Apart from that I am not sure asking nm to demangle is correct.

https://github.com/mesonbuild/meson/blob/4f808400371cb70e76a4970e59b64a9ea7663caf/mesonbuild/backend/ninjabackend.py#L3696

You mean to say something along the lines of the above link. If yes, then it is a nice idea. That way we can reuse the code. I will implement and adapt. We need not to use `/usr/bin/ld` rule then like in this version of the PR.

Regarding demangling, yes, let me remove the -C option. Will use only the B-BSD format and g to take global symbols. 

Thank you very much for the input to make this draft look better.

bonzini commented:
More changes suggested [here](https://github.com/mesonbuild/meson/commit/d2598a7309467bce57f2c3beddc67d3cdf014704#r183800244):
* try to place the `generate_prelink` where the existing call is, so that you can change `final_obj_list` to include only the prelinked object.
* not sure if the change around `if prelinker.get_prelink_append_compile_args():` is needed, it could be a configuration error
* use `set()` to initialize `result` and avoid adding `#!` that is later removed.

KamathForAIX commented:
> * not sure if the change around `if prelinker.get_prelink_append_compile_args():` is needed, it could be a configuration error

@bonzini 

I can tell why I did that.

So in a real-world scenario I am testing this with SciPy, glib2 and PostgreSQL master branches in AIX.

Gilb2 is built in both 32- and 64-bit modes.

So if we do not do that and are compiling in 64-bit mode,

`gcc -r -o src/interfaces/libpq/libpq.so.5.19.p/libpq-prelink.o src/interfaces/libpq/libpq.so.5.19.p/fe-auth-oauth.c.o src/interfaces/libpq/libpq.so.5.19.p/fe-auth-scram.c.o `

fe-auth-scram.c.o, fe-auth-oauth.c.o are 64-bit mode objects, but the default compilation mode in AIX is 32 bit. So, libpq-prelink.o will become 32-bit, which the linker is complaining about. 

Any thoughts you have? I suspect the CIs on other targets are failing because of that, which I will inspect as well because everything else is now guarded for AIX.

So we need to pass -maix64 or -maix32 which is compiler args.

bonzini commented:
How are you normally ensuring that `-maix64` is passed? "Multilib flags" like `-maix64` and `-maix32` should be part of `CC` (or `c` in the `[binaries]` section of the machine file), not `CFLAGS`/`c_args`.

KamathForAIX commented:
@bonzini 

Right now we expect users will give it via CC or CFLAGS but have an open mind to do it other ways if you know how to make this better or strict.

bonzini commented:
> Right now we expect users will give it via CC or CFLAGS but have an open mind to do it other ways

CC should work without that change, while CFLAGS will have problems with Meson in general.

Also, [IBM docs](https://www.ibm.com/docs/en/xl-c-aix/13.1.0?topic=library-compiling-shared) say that "if you do not specify a -bE option, all the global symbols are exported except for those symbols that have the hidden or internal visibility attribute". What is the case in which the new code helps?

Finally, I would like more information on this:

```
        # Do not export internal symbols like _GLOBAL* or start with __
        if symbol_name.startswith('_GLOBAL') or symbol_name.startswith('__'):
            continue
```

Are these AIX-isms? User-defined symbols should be included in the shared libraries even if they start with "'__'" for example.

KamathForAIX commented:
> > Right now we expect users will give it via CC or CFLAGS but have an open mind to do it other ways
> 
> CC should work without that change, while CFLAGS will have problems with Meson in general.
> 

Sure, then I am going to remove that change in the coming push. We will rely on CC.

> Also, [IBM docs](https://www.ibm.com/docs/en/xl-c-aix/13.1.0?topic=library-compiling-shared) say that "if you do not specify a -bE option, all the global symbols are exported except for those symbols that have the hidden or internal visibility attribute". What is the case in which the new code helps?

I can give you 3 good reasons why we do that.

1: Stable ABI: If we do not have export files, the set of exported symbols can change just from rebuilding [new compilers or helper functions]. So users who made a library without an export file may rely on an internal symbol exported unintentionally that will cause runtime 'symbols not defined' issues.

2: Symbol leakage: We do not want to leak undefined symbols marked as U. [See in nm output below]. This can create painful issues if users actually want to actually declare and define creating conflicts.

3: Toolchain consistency with CMake and LLVM. 
At IBM we have 3 compilers: GCC, Clang, and IBM-OpenXL (which is built on Clang).
GCC does not support LTO.
But the other two do on AIX. And what LTO objects export by default is not consistent and can get hard to debug later on.

If I can say in one line, we want to prevent ABI rift and symbol leakage and unwanted conflicts.

CMake has this behaviour already in AIX. In fact, we also have an llvm-nm to extract symbols from LTO objects.

> Finally, I would like more information on this:
> 
> ```
>         # Do not export internal symbols like _GLOBAL* or start with __
>         if symbol_name.startswith('_GLOBAL') or symbol_name.startswith('__'):
>             continue
> ```
> 


> Are these AIX-isms? User-defined symbols should be included in the shared libraries even if they start with "'__'" for example.

Yes. Everyone is curious of this in AIX :) 

See the output below. _GLOBAL__AIXD_libmylib_so are all compiler generator constructors and destructors that I do not desire to export. These get exported, and then tomorrow I change my compiler to Clang; these symbols will not be there. So if I update this library, then I will have to update all libraries/binaries linking to the ./build/libmylib.a which is not a desirable thing since we have to maintain backward compatibility in AIX.  

This also will answer your second question. 

Hope I was able to explain. 

```
# nm -BCg  ./build/libmylib.a

libmylib.so.1.0.0:
100004d8 T ._GLOBAL__AIXD_libmylib_so
10000474 T ._GLOBAL__AIXI_libmylib_so
100002d0 T .global destructors keyed to 65535_0___dso_handle
100003a8 T ._GLOBAL__FD_libmylib_so
100002e0 T ._GLOBAL__FI_libmylib_so
10000268 T .__cxa_finalize
100002a0 T .__init_aix_libgcc_cxa_atexit
1000053c T .__strtollmax
1000061c T .add_numbers
10000674 T .multiply_numbers
100006d0 T .print_message
10000568 T .printf
1000078c T .process_value
200008e8 D _GLOBAL__AIXD_libmylib_so
         U _GLOBAL__AIXD_shr_o
200008dc D _GLOBAL__AIXI_libmylib_so
         U _GLOBAL__AIXI_shr_o
200008b8 D global destructors keyed to 65535_0___dso_handle
200008d0 D _GLOBAL__FD_libmylib_so
200008c4 D _GLOBAL__FI_libmylib_so
         U __cxa_finalize
20000820 D __dso_handle
         U __strtollmax
200008f4 D add_numbers
20000958 B buffer
20000950 B counter
20000840 D library_version
20000900 D multiply_numbers
2000090c D print_message
         U printf
20000918 D process_value
```

bonzini commented:
This makes sense, and I agree that it's a good reason to have a default handling of -bE, unlike what we do with other dynamic linkers.

However your handling of '.' symbols seems wrong. They are always of T type so they would override everything that you do in aix_syms. Either you never want them public or you always want them (I don't know—never seems more logical), but I don't think it's correct to apply the usual logic to these '.' symbols and use the result for the non-prefixed symbol.

For `__`, I would rather leave it out. If there are any AIX-specific symbols that should always be discarded you can special case them, but even then I would think twice about it.

I know @eli-schwartz wants to have generic symbol list functionality. We will have to sort out the order in which to commit these two pieces of functionality but AIX does seem special in this respect, which is what I wanted to ascertain with this review.

eli-schwartz commented:
See also  #14980

bonzini commented:
@KamathForAIX another question, what happens if you have multiple `-bE:` options?  Are they summed, or does the last one win?

KamathForAIX commented:
> @KamathForAIX another question, what happens if you have multiple `-bE:` options? Are they summed, or does the last one win?

@bonzini AIX in this case, combines both the export file symbols and exports. If there are duplicates, the linker will only give warning since it wants to tell the project build that you may have two functions which may or may not be of the same implementation. To confirm this I have tried with an example.

Consider the shared library mylib.a
```
# cat mylib.c
#include <stdio.h>

void function_one(void) {
    printf("Function One called\n");
}

void function_two(void) {
    printf("Function Two called\n");
}

void function_three(void) {
    printf("Function Three called\n");
}

void internal_helper(void) {
    printf("Internal helper (should not be exported)\n");
}
```

Consider the code main that uses this library,
```
# cat main.c
#include <stdio.h>

// Declare the functions from the shared library
extern void function_one(void);
extern void function_two(void);
extern void function_three(void);

int main(void) {
    printf("Testing multiple -bE: export files\n");
    printf("===================================\n\n");
    
    function_one();
    function_two();
    function_three();
    
    printf("\nAll functions called successfully!\n");
    return 0;
}
```

Now consider two export files,
```
# cat export1.exp 
#!
function_one

# cat export2.exp 
#!
function_two
function_three
```

Let us now compile.
```
gcc -c -fPIC mylib.c -o mylib.o
gcc -shared -o libmylib.so mylib.o -Wl,-bE:export1.exp -Wl,-bE:export2.exp
gcc -c main.c -o main.o
gcc -o test_program main.o -L. -lmylib -Wl,-blibpath:$(pwd):/usr/lib
```
After compiling if we check the dump of the shared library we get,
```
# dump -Tv libmylib.so | grep function
[8]     0x20000798    .data      EXP     DS SECdef        [noIMid] function_one
[9]     0x200007a4    .data      EXP     DS SECdef        [noIMid] function_two
[10]    0x200007b0    .data      EXP     DS SECdef        [noIMid] function_three
```
Clearly, all three functions are exported.

KamathForAIX commented:
@bonzini @eli-schwartz 

Thank you for your patience. I was reading that issue and researching around for more on this topic. So here is my study.

Yes, I can confirm that in Power Linux, in certain projects, they use -Wl,--version-script and pass a map file which is an export file only, like in Power AIX. Their map file looks something like below. 

```
{
  global:
    function_one;
    function_two;
  local:
    *;
};
```

But one of the places where this feature will be super useful is PostgreSQL. It is one of the nicest things to try. 

Since database workloads need to make sure that compiler symbols are not exported to maintain compatibility.

If we see the code [here](https://github.com/postgres/postgres/blob/master/meson.build#L236) they do the same. Just that it is in their own custom meson.build file via custom commands. If I understood that correctly, Windows and Darwin also do similar things.

Having said that, I would like to work on another draft PR if we can agree, as @eli-schwartz pointed out in issue #14980 .

The issue for that PR is if a project decides to create its own export file instead of us creating it like in this draft PR, we need to respect that. For example, again, PostgreSQL does that [here](https://github.com/postgres/postgres/blob/574581b50ac9c63dd9e4abebb731a3b67e5b50f6/src/interfaces/libpq-oauth/exports.txt#L2). 

It is a very interesting concept. 
Thank you @bonzini  for asking that question about two export files. That is what prompted me to research. Let me know what you all think. 

To summarise, we can have two commits:
1: If there is no export file, then Meson extracts symbols via the symbol extractor and creates one.
2: If a project does use an export file like PostgreSQL does, then we will respect that, use that and then not go on to create the export file ourselves.

bonzini commented:
(Wrong message before - you did answer it).

I think we need to bite the bullet and implement the famous "compiler options IR". This is the easiest way to make AIX respect the user's -bE option while being able to add its own.

KamathForAIX commented:
Ah yes. We can do that. Now I get why you asked that question. That makes sense.

KamathForAIX commented:
I suspect this `elif t.aix_so_archive` in get_dependency_filename () is the root cause of all CI failures since everywhere else changes are guarded for AIX and Linux might be expecting `self.get_target_shsym_filename(t)` . Let me fix this with a push and then go about my work to implement the IR. I will guard it with self.environment.machines[t.for_machine].is_aix() so that it sets right since aix_so_archive is true for all shared library instances.

bonzini commented:
There is already some initial work on the IR but the complex part is the design. For now please leave this aside. If anything, you could start looking at generic support for a symbol list file (which would be translated into one of ld version script, Windows .def file or AIX export list).

KamathForAIX commented:
> If anything, you could start looking at generic support for a symbol list file (which would be translated into one of ld version script, Windows .def file or AIX export list).

Sure @bonzini 

I can help with that. Hopefully the CIs will be green now.

KamathForAIX commented:
> However your handling of '.' symbols seems wrong. They are always of T type so they would override everything that you do in aix_syms. Either you never want them public or you always want them (I don't know—never seems more logical), but I don't think it's correct to apply the usual logic to these '.' symbols and use the result for the non-prefixed symbol.
> 
> For `__`, I would rather leave it out. If there are any AIX-specific symbols that should always be discarded you can special case them, but even then I would think twice about it.

This is right. Let me remove the '.' symbol processing logic and test this. Thanks.

KamathForAIX committed:
Teach meson to handle a custom export file if given

This patch helps meson only export symbols mentioned in a export file for a shared library in AIX.

KamathForAIX commented:
@bonzini @eli-schwartz 

> There is already some initial work on the IR but the complex part is the design. For now please leave this aside. If anything, you could start looking at generic support for a symbol list file (which would be translated into one of ld version script, Windows .def file or AIX export list).

I am pushing commit number 2 as a draft, which respects an export symbol list in AIX, and other targets can follow. Leaving the IR aside, just simple changes to handle an export file if it is used in AIX. The idea is if we look at the two compilation commands below, when there is an export file, it goes into the symbol extractor. Otherwise, the symbol extractor has 4 arguments.

I have made it simple as of now. We should be able to use it like the below example. [See symbol_list argument].  An export file, if passed, means we use it to pass as -Wl,bE or -Wl,--version-script. And only symbols in the export file are exported. Otherwise, we will extract the symbols from the object files used to create the shared library, like the commit 1.  

```
# Create shared library with build_rpath
mylib = shared_library('mylib',
  'mylib.c',
  symbol_list : 'exports.txt',
  install : true,
  install_dir : get_option('libdir'),
  build_rpath : '/meson_test/build:/tmp/fake_build_path',
  install_rpath : '$ORIGIN/rad/../lib')
```

After this we get,
```
[4/7] /meson/install/opt/freeware/bin/meson --internal symbolextractor /meson_test/build libmylib.so.p/mylib-prelink.o libmylib.a libmylib.so.p/libmylib.so.symbols  /meson_test/exports.txt
[5/7] gcc -maix64  -o libmylib.so libmylib.so.p/mylib-prelink.o -Wl,-bernotok -Wl,-bnoipath -Wl,-bbigtoc -shared -fPIC -Wl,-bE:/meson_test/build/libmylib.so.p/libmylib.so.symbols '-Wl,-blibpath:$ORIGIN/rad/../lib:/meson_test/build:/tmp/fake_build_path:/opt/freeware/lib/gcc/powerpc-ibm-aix7.2.0.0/13/ppc64:/opt/freeware/lib/ppc64:/opt/freeware/lib/gcc/powerpc-ibm-aix7.2.0.0/13:/opt/freeware/lib:/lib:/usr/lib'
```

If we do not use an export file we get,

```
[4/7] /meson/install/opt/freeware/bin/meson --internal symbolextractor /meson_test/build libmylib.so.p/mylib-prelink.o libmylib.a libmylib.so.p/libmylib.so.symbols  
[5/7] gcc -maix64  -o libmylib.so libmylib.so.p/mylib-prelink.o -Wl,-bernotok -Wl,-bnoipath -Wl,-bbigtoc -shared -fPIC -Wl,-bE:/meson_test/build/libmylib.so.p/libmylib.so.symbols '-Wl,-blibpath:$ORIGIN/rad/../lib:/meson_test/build:/tmp/fake_build_path:/opt/freeware/lib/gcc/powerpc-ibm-aix7.2.0.0/13/ppc64:/opt/freeware/lib/ppc64:/opt/freeware/lib/gcc/powerpc-ibm-aix7.2.0.0/13:/opt/freeware/lib:/lib:/usr/lib'
```

In AIX the only processing we do in aix_extract_sym_from_file() is to remove comments and empty lines. Linux might want to write their own logic in their --version-script format. Other targets may also want to write their own **exported_symbols_args** indicating whether it is -bE or --version-script and so on.

Let me know what you think. We can redesign as well. 

Looking at the PostgreSQL export files, I am assuming all targets will have an export list like the below:
```
# cat  exports.txt
# Export file
hello_from_lib
hello
```

#14980 mentions export_symbols, and they want to do it for shared modules, shared libraries and executables.

But as of now I am restricted to just the shared library and should be okay to expand once we all agree. 

Also, export lists will be placed in the source directory and not outside. These are assumptions I made, but feel free to tell me if I missed something or we can make this better.

Thank you for all your guidance so far.

TheOneric commented on mesonbuild/scripts/symbolextractor.py:
```diff
@@ -383,18 +383,54 @@ def gen_symbols(libfilename: str, impfilename: str, outfilename: str, cross_host
+                line = line.strip()
+                # Skip comments that start with #
+                # Skip empty lines also.
+                if not line and line.startswith('#'):
```
Probably should be `not line or line.startswith('#')`

KamathForAIX replied:
Ah yes!

Thank you for pointing out. Let me test and update.

KamathForAIX replied:
Thanks @TheOneric 

Have updated the same.

KamathForAIX replied:
Resolving merge conflicts and test the same.

KamathForAIX commented:
I have resolved the merge conflicts and made the PR working for this feature.

By the way, if any maintainers need an AIX LPAR for Meson, then let me know. IBM provides the same via OSU Labs, and I can get them for you. Mentioned it [here](https://github.com/mesonbuild/meson/issues/15831) as well. Thanks.
